### PR TITLE
atomics: switch to using Pair for return pairs

### PR DESCRIPTION
**Pull request by vtjnash**
This makes many more functions type-stable, by directly preserving the
element type when making the copy, and prints as `old => new`, like the
argument to atomic replace.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
I'm not totally sold on this; the case of `@atomicreplace` undermines it a bit since the old value and success flag are really just separate return values, not related as a replacement mapping or key-value pair.

**Comment by vtjnash** [comment]
That particular usage of pair is a common API pattern in C++ to return `old => success`, e.g.  in https://en.cppreference.com/w/cpp/container/map/insert 😂

**Comment by JeffBezanson** [comment]
That's a good example. In C++ pair is basically the same as a 2-tuple, though. I guess they have tuples now, but pair predates them by a lot? It looks to me like they'd use 2-tuples now if they were doing it over.

**Comment by JeffBezanson** [comment]
Or, if we want something invariant, another possibility is to use namedtuples, and be more self-documenting at the same time.

File: base/boot.jl [hunk 1]
```diff
@@ -813,4 +813,16 @@ _parse = nothing
 # support for deprecated uses of internal _apply function
 _apply(x...) = Core._apply_iterate(Main.Base.iterate, x...)
 
+struct Pair{A, B}
+    first::A
+    second::B
+    # if we didn't inline this, it's probably because the callsite was actually dynamic
+    # to avoid potentially compiling many copies of this, we mark the arguments with `@nospecialize`
+    # but also mark the whole function with `@inline` to ensure we will inline it whenever possible
+    # (even if `convert(::Type{A}, a::A)` for some reason was expensive)
```
Comment (martinholters) [inline base/boot.jl:822]: Should this comment have been moved to Base.jl? Or which method does it belong to?

**Pull request merged**